pkg/controller/encryption/ipsec/ipsec_handler.go

```diff
@@ -0,0 +1,481 @@
+/*
+ * Copyright The Kmesh Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ipsec
+
+import (
+	"bufio"
+	"crypto/sha512"
+	"encoding/json"
+	"fmt"
+	"hash/fnv"
+	"net"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/vishvananda/netlink"
+	"istio.io/istio/pkg/filewatcher"
+)
+
+const (
+	IpSecKeyFile  = "./kmesh-ipsec/keys"
+	offsetSpi     = 0
+	offsetAead    = 1
+	offsetAeadKey = 2
+	offsetAeadLen = 3
+)
+
+type ipSecInfo struct {
+	bootID        string
+	spi           int8
+	nodeID        uint32
+	spiCreateTime string
+}
+
+type IpSecKey struct {
+	Spi         int8   `json:"spi"`
+	AeadKeyName string `json:"aeadKeyName"`
+	AeadKey     []byte `json:"aeadKey"`
+	Length      int    `json:"length"`
+	CreateTime  string `json:"createTime"`
+}
+
+type IpSecHandler struct {
+	IpSecKey
+	mutex       sync.RWMutex
+	watcher     filewatcher.FileWatcher
+	nodeInfos   map[string]ipSecInfo
+	oldIpSecKey map[string]IpSecKey
+}
+
+func NewIpSecHandler() *IpSecHandler {
+	return &IpSecHandler{
+		nodeInfos:   make(map[string]ipSecInfo),
+		oldIpSecKey: make(map[string]IpSecKey),
+	}
+}
+
+func (is *IpSecHandler) LoadIPSecKeyFromFile(filePath string) error {
+	if is.IpSecKey.Spi != 0 {
+		is.oldIpSecKey[is.CreateTime] = is.IpSecKey
+	}
+
+	for key := range is.oldIpSecKey {
+		keyCreateTime, err := time.Parse(TimeFormatString, key)
+		if err != nil {
+			return fmt.Errorf("failed to parser key create time, %v", err)
+		}
+		if keyCreateTime.Add(1 * time.Hour).Before(time.Now()) {
+			delete(is.oldIpSecKey, key)
```

Comment:
SHould we check all other nodes' spi rather than depending on the time to delete a old key